rfcs/convergence/react-focus-management.md

```diff
@@ -0,0 +1,169 @@
+| focusable | FocusableAPI | API for traversing focusable elements |
+
+The below table defines the props used by the provider component. All context values can be used as props in the provider component but are optional.
+
+| Name                   | Value              | Description                                                                                                           |
+| ---------------------- | ------------------ | --------------------------------------------------------------------------------------------------------------------- |
+| dir                    | 'ltr' \| 'rtl'     | Direction indicator, some keyboard behaviour will need to change based on this value                                  |
+| Deloser                | DeloserAPI         | Tools to help automatically restore focus when it gets lost without manually calling `.focus()` from application code |
+| internal\_\_ahInstance | AbilityHelpersCore | The ability helpers instance, not expected to be be used directly                                                     |
+
+#### Hooks
+
+This spec proposes an API based on React hooks pattern. [ability-helpers](https://github.com/microsoft/ability-helpers) works mainly on adding `data-*` attributes to DOM elements to apply configurable behaviours to the 'groups' of focusables.
+
+However some internal behaviours such as `modalizer` or `deloser` require DOM elements during the configuration phase, which will be most easily configured in React using `refs`.
+
+Therefore it must be noted that **although some hooks might not require ref interaction and can simply return attributes to be spread, in an effort to maintain consistency refs are used, which fits the Fluent component design pattern which enforces ref forwarding**
+
+```typescript
+// Basic up/down keyboard navigation
+const ref = useKeyboardNavigationGroup();
+
+<div ref={ref}>{props.children}</div>;
+```
+
+```typescript
+const ref = React.useReact<HTMLElement>();
+const {
+  findFirst,
+  findLast,
+  findAll,
+} = useFocusable(ref); // All operations are with respect to the given DOM container
+
+findFirst() // first focusable
+findLast() // first last focusable
+findAll(condition: () => true) // first all focusables with condition
+
+<div ref={ref}>{props.children}</div>
+```
+
+#### useFocusable
+
+A hook that exposes helper methods to traverse focusable elements in the context of a DOM element.
+
+The below table lists the helper functions to support:
+
+| Name      | Value         | Options                  | Description                                         |
+| --------- | ------------- | ------------------------ | --------------------------------------------------- |
+| findFirst | HTMLElement   |                          | Finds the first focusable element                   |
+| findLast  | HTMLElement   |                          | Finds the last focusable element                    |
```

Comment:
Just wondering. Will we have `findPrevious` and `findNext`?

Reply:
AH does expose them, but I'm not sure of a use case yet where they will be necessary, so I would rather iterate and add them when we have a use case for it